app/bundles/PointBundle/Model/PointModel.php

```diff
@@ -234,77 +242,46 @@ public function triggerAction($type, $eventDetails = null, $typeId = null, Lead
+            if (isset($settings['eventName'])) {
+                // 1. step - can change points from event
+                $pointChangeActionExecutedEvent = new PointChangeActionExecutedEvent($action, $lead, $eventDetails, $completedActions);
+                $event                          = $this->dispatcher->dispatch($settings['eventName'], $pointChangeActionExecutedEvent);
+                if (!$event->canChangePoints()) {
+                    continue;
                 }
-
-                $pass = [];
-                foreach ($reflection->getParameters() as $param) {
-                    if (isset($args[$param->getName()])) {
-                        $pass[] = $args[$param->getName()];
-                    } else {
-                        $pass[] = null;
-                    }
+            } elseif (!$action->getRepeatable()) {
+                // 2. step - can change points from callback
+                if (!$this->invokeCallback($action, $lead, $eventDetails, $settings)) {
+                    continue;
                 }
-                $pointsChange = $reflection->invokeArgs($this, $pass);
-
-                if ($pointsChange) {
-                    $delta = $action->getDelta();
-                    $lead->adjustPoints($delta);
-                    $parsed = explode('.', $action->getType());
-                    $lead->addPointsChangeLogEntry(
-                        $parsed[0],
-                        $action->getId().': '.$action->getName(),
-                        $parsed[1],
-                        $delta,
-                        $ipAddress
-                    );
-
-                    $event = new PointActionEvent($action, $lead);
-                    $this->dispatcher->dispatch(PointEvents::POINT_ON_ACTION, $event);
-
-                    if (!$action->getRepeatable()) {
-                        $log = new LeadPointLog();
-                        $log->setIpAddress($ipAddress);
-                        $log->setPoint($action);
-                        $log->setLead($lead);
-                        $log->setDateFired(new \DateTime());
-                        $persist[] = $log;
-                    }
+                // 3. step - can change points from log
+                if (isset($completedActions[$action->getId()])) {
+                    continue;
                 }
             }
+
+            $this->adjustLeadPoints($action, $lead);
+
+            $event = new PointActionEvent($action, $lead);
+            $this->dispatcher->dispatch(PointEvents::POINT_ON_ACTION, $event);
+
+            // Add to log, repeatable is not logged, just executed
+            if (!$action->getRepeatable()) {
+                $log = new LeadPointLog();
+                $log->setIpAddress($ipAddress);
+                $log->setPoint($action);
+                $log->setLead($lead);
+                $log->setInternalId($this->propertyAccessor->getValue($eventDetails, 'id'));
```

Comment:
@kuzmany question about this. Can we set it for int only? I know you mention this in the description:

> added new column internal_id to point_lead_action_log table - ID or string - in thu future we can store URL for page hit and then we can run points action on each visits

But we cannot store URL to 191 characters. So I think we should go with int for now as that's what we use it for now.

Reply:
Wait, what is this ID used for again? It's big part of this PR but I don't see any reason for it.

Reply:
internal_id is set for any ID of item. At the moment it use for store email_id to check If was point trigger on selected emails. 

String we assume support for plugins, If we need check some string ID of external service item.

Reply:
I see. It would be quite confusing to store external ID in a column called internal_id. I think it deserves another column. Or a plugin can create another table to keep whatever data it needs without need for a core table change.

What about the change I did to remove it from the primary key columns and making it nullable. Will it break anything? Why was it primary?

Reply:
It was build for M2 times ago, then  yes, naming is probably confusing and some code is not the best written.
That column is  something like integration_object_id in sync_object_mapping table, also is varchar. 

Before point action was triggered just for point_id/lead_id one time, this option allow point action trigger in combination point_id/lead_id/object_id. Then you are able to trigger one point action for selected forms, pages etc etc.

Reply:
Primary key does not make sense on it, of course. Just need index